src/common/src/util/value_encoding/mod.rs

```diff
@@ -53,6 +54,20 @@ pub fn serialize_datum_ref(datum_ref: &DatumRef<'_>, buf: &mut impl BufMut) {
     }
 }
 
+/// Serialize a `ScalarImpl` into bytes.
+pub fn serialize_scalar_impl(scalar: &ScalarImpl) -> Vec<u8> {
+    let mut bytes = vec![];
+    serialize_value(scalar.as_scalar_ref_impl(), &mut bytes);
+    bytes
+}
+
+pub fn deserialize_scalar_impl_from_prost(
+    prost: &ProstDatum,
+    data_type: &DataType,
+) -> Result<ScalarImpl> {
+    deserialize_value(data_type, &mut &*prost.body)
```

Comment:
This semantic seems incorrect. The datum has a null tag prepended to the scalar.

Reply:
You are right. :(